File: tree/102. Binary Tree Level Order Traversal.py
"""
这种题目，套路现在很熟练了。但是，一个新思路，如何避免使用两个queue呢？使用一个size 变量就可以了。
"""
class Solution:
    """
    @param root: A Tree
    @return: Level order a list of lists of integer
    """
    def levelOrder(self, root):
        # write your code here
        res = []
        if root:
            q = collections.deque([root])
            size = 1
            while q:
                temp = []
                i = 0
                new_size = 0
                while i < size:
                    node = q.popleft()
                    temp.append(node.val)
                    if node.left:
                        q.append(node.left)
                        new_size += 1
                    if node.right:
                        q.append(node.right)
                        new_size += 1
                    i += 1
                size = new_size
                res.append(temp)
        return res

File: tree/102. Binary Tree Level Order Traversal.c
// 用C写这两段程序是非常不容易的。之前犯了好几个错误。
// 1. 没有initialize returnSize。
// 2. 没有给所有的分配内存。
// 3. 不是特别会用malloc。


void traverse(struct TreeNode* node, int** res, int* returnColumnSizes, int depth, int* returnSize) {
    if(!node)
        return;
    *returnSize = (depth + 1 > *returnSize) ? depth + 1 : *returnSize;
    if (!returnColumnSizes[depth]) {
        res[depth] = malloc(sizeof(int)*2001);
      }
    res[depth][returnColumnSizes[depth]++] = node->val;
    traverse(node->left, res, returnColumnSizes, depth + 1, returnSize);
    traverse(node->right, res, returnColumnSizes, depth + 1, returnSize);
}

int** levelOrder(struct TreeNode* root, int* returnSize, int** returnColumnSizes){
    int** res = malloc(sizeof(int*) * 2001);
    *returnColumnSizes = malloc(sizeof(int) * 2001);
    *returnSize = 0;
    for(int i=0; i < 2001; i++)
        (*returnColumnSizes)[i] = 0;
    traverse(root, res, *returnColumnSizes, 0, returnSize);
    return res;    
}
